File: backend/src/main/java/com/crimemanagement/controller/CrimeController.java
package com.crimemanagement.controller;

import com.crimemanagement.model.Crime;
import com.crimemanagement.service.CrimeService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.time.LocalDateTime;
import java.util.List;

@RestController
@RequestMapping("/api/crimes")
@CrossOrigin(origins = "*")
public class CrimeController {
    
    @Autowired
    private CrimeService crimeService;
  
    @PostMapping
    public ResponseEntity<Crime> createCrime(@RequestBody Crime crime) {
        // Log the incoming crime object for debugging
        System.out.println("Received crime for creation: " + crime.getCrimeType() + " at " + crime.getLocation());
        if (crime.getImages() != null && !crime.getImages().isEmpty()) {
            System.out.println("Received " + crime.getImages().size() + " images (Base64 strings).");
        }

        // Set server-side properties that are not sent from client or need to be set on creation
        crime.setDateTime(LocalDateTime.now());
        crime.setStatus("REPORTED");
        crime.setCriminalStatus("UNKNOWN"); // Set default criminal status on creation

        // Handle Serial Killer specific properties based on crimeType
        if ("Serial Killer".equalsIgnoreCase(crime.getCrimeType())) {
            crime.setPriority("HIGH");
            crime.setAlertLevel("CRITICAL");
            crime.setSpecialUnit("Criminal Investigation Department");
            crime.setRequiresImmediateAttention(true);
        } else {
            crime.setPriority("MEDIUM");
            crime.setAlertLevel("NORMAL");
            crime.setRequiresImmediateAttention(false);
        }

        try {
            Crime savedCrime = crimeService.saveCrime(crime);
            System.out.println("Crime saved successfully with ID: " + savedCrime.getId());
            return ResponseEntity.ok(savedCrime);
        } catch (Exception e) {
            System.err.println("Error saving crime: " + e.getMessage());
            e.printStackTrace(); // Print full stack trace for detailed error
            return ResponseEntity.status(500).build(); // Return 500 Internal Server Error
        }
    }
    
    @GetMapping
    public ResponseEntity<List<Crime>> getAllCrimes() {
        List<Crime> crimes = crimeService.getAllCrimes();
        return ResponseEntity.ok(crimes);
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<Crime> getCrimeById(@PathVariable String id) {
        Crime crime = crimeService.getCrimeById(id);
        if (crime != null) {
            return ResponseEntity.ok(crime);
        }
        return ResponseEntity.notFound().build();
    }
    
    @GetMapping("/type/{crimeType}")
    public ResponseEntity<List<Crime>> getCrimesByType(@PathVariable String crimeType) {
        List<Crime> crimes = crimeService.getCrimesByType(crimeType);
        return ResponseEntity.ok(crimes);
    }
    
    @GetMapping("/location/{location}")
    public ResponseEntity<List<Crime>> getCrimesByLocation(@PathVariable String location) {
        List<Crime> crimes = crimeService.getCrimesByLocation(location);
        return ResponseEntity.ok(crimes);
    }
    
    @GetMapping("/search")
    public ResponseEntity<List<Crime>> searchCrimes(@RequestParam String query) {
        List<Crime> crimes = crimeService.searchCrimes(query);
        return ResponseEntity.ok(crimes);
    }
    
    @PutMapping("/{id}")
    public ResponseEntity<Crime> updateCrime(@PathVariable String id, @RequestBody Crime crime) {
        // Log the incoming crime object for debugging
        System.out.println("Received crime for update (ID: " + id + "): " + crime.getCrimeType() + " Status: " + crime.getStatus() + " Criminal Status: " + crime.getCriminalStatus());

        if (crimeService.getCrimeById(id) != null) { // Check if crime exists
            crime.setId(id); // Ensure the ID is set for update
            // The saveCrime method will handle updating the existing record
            Crime updatedCrime = crimeService.saveCrime(crime); 
            System.out.println("Crime updated successfully with ID: " + updatedCrime.getId());
            return ResponseEntity.ok(updatedCrime);
        }
        System.err.println("Crime with ID " + id + " not found for update.");
        return ResponseEntity.notFound().build();
    }
    
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteCrime(@PathVariable String id) {
        crimeService.deleteCrime(id);
        return ResponseEntity.noContent().build();
    }
}

File: controllers/crime.controller.js
const express = require("express")
const crimeService = require("../services/crime.service")
const router = express.Router()

router.post("/", async (req, res) => {
  const crime = req.body
  console.log("Received crime for creation:", crime.crimeType, "at", crime.location)
  if (crime.images && crime.images.length > 0) {
    console.log("Received", crime.images.length, "images (Base64 strings).")
  }

  // Set server-side properties that are not sent from client or need to be set on creation
  crime.dateTime = new Date() // Equivalent to LocalDateTime.now()
  crime.status = "REPORTED"
  crime.criminalStatus = "UNKNOWN" // Set default criminal status on creation

  // Handle Serial Killer specific properties based on crimeType
  if ("Serial Killer".equalsIgnoreCase(crime.crimeType)) {
    crime.priority = "HIGH"
    crime.alertLevel = "CRITICAL"
    crime.specialUnit = "Criminal Investigation Department"
    crime.requiresImmediateAttention = true
  } else {
    crime.priority = "MEDIUM"
    crime.alertLevel = "NORMAL"
    crime.requiresImmediateAttention = false
  }

  try {
    const savedCrime = await crimeService.saveCrime(crime)
    console.log("Crime saved successfully with ID:", savedCrime._id)
    return res.status(200).json(savedCrime)
  } catch (e) {
    console.error("Error saving crime:", e.message)
    console.error(e.stack) // Print full stack trace for detailed error
    return res.status(500).send("Internal Server Error")
  }
})

router.get("/", async (req, res) => {
  const crimes = await crimeService.getAllCrimes()
  return res.status(200).json(crimes)
})

router.get("/:id", async (req, res) => {
  const { id } = req.params
  const crime = await crimeService.getCrimeById(id)
  if (crime) {
    return res.status(200).json(crime)
  }
  return res.status(404).send("Crime not found")
})

router.get("/type/:crimeType", async (req, res) => {
  const { crimeType } = req.params
  const crimes = await crimeService.getCrimesByType(crimeType)
  return res.status(200).json(crimes)
})

router.get("/location/:location", async (req, res) => {
  const { location } = req.params
  const crimes = await crimeService.getCrimesByLocation(location)
  return res.status(200).json(crimes)
})

router.get("/search", async (req, res) => {
  const { query } = req.query
  const crimes = await crimeService.searchCrimes(query)
  return res.status(200).json(crimes)
})

router.put("/:id", async (req, res) => {
  const { id } = req.params
  const crimeData = req.body
  console.log(
    "Received crime for update (ID:",
    id + "):",
    crimeData.crimeType,
    "Status:",
    crimeData.status,
    "Criminal Status:",
    crimeData.criminalStatus,
  )

  const existingCrime = await crimeService.getCrimeById(id)
  if (existingCrime) {
    crimeData.id = id // Ensure the ID is set for update
    const updatedCrime = await crimeService.saveCrime(crimeData) // saveCrime handles update if ID is present
    console.log("Crime updated successfully with ID:", updatedCrime._id)
    return res.status(200).json(updatedCrime)
  }
  console.error("Crime with ID", id, "not found for update.")
  return res.status(404).send("Crime not found for update")
})

router.delete("/:id", async (req, res) => {
  const { id } = req.params
  await crimeService.deleteCrime(id)
  return res.status(204).send() // No Content
})

module.exports = router
